R/wqp_getcounts.R

```diff
@@ -0,0 +1,97 @@
+# getting counts of sites and records for each characteristic group, site type, and year
+
+library(dataRetrieval)
+library(httr)
+
+char_types <- c('Physical', 'Inorganics, Major, Metals', 'Inorganics, Major, Non-metals', 
+                'Inorganics, Minor, Metals', 'Inorganics, Minor, Non-metals', 'Not Assigned',
+                'Nutrient', 'Organics, Other', 'Organics, PCBs', 'Organics, Pesticide', 
+                'Microbiological', 'Biological', 'Information', 'Sediment', 'Radiochemical', 
+                'Stable Isotopes', 'Population/Community', 'Toxicity')
+
+site_types <- c('Aggregate groundwater use', 'Aggregate surface-water-use', 'Atmosphere', 'Estuary', 
+                'Facility', 'Glacier', 'Lake, Reservoir, Impoundment', 'Land', 'Not Assigned', 
+                'Ocean', 'Spring', 'Stream', 'Subsurface', 'Well', 'Wetland')
+
+start_dates <- seq(as.Date('1800-01-01'), 
+                   as.Date(paste0(format(Sys.Date(), "%Y"), '-01-01')),
+                   by = 'year')
+add_yr <- seq(tail(start_dates,1), length = 2, by = '1 year')[2]
+end_dates <- c(start_dates[-1], add_yr) -1 # end_dates are December 31st of following yr
+dates_df <- data.frame(start_date = start_dates, end_date = end_dates, stringsAsFactors = FALSE)
+
+createCountDF <- function(char_type, site_type, start_date, end_date){
+  #format dates for WQP call
+  start_date <- format(start_date, format = '%m-%d-%Y')
+  end_date <- format(end_date, format = '%m-%d-%Y')
+  
+  d <- retryWQP(characteristicType=char_type, siteType = site_type,
+                startDate = start_date, endDate = end_date, querySummary = TRUE,
+                retries = 5)
+  
+  if(is.null(d$`total-site-count`)){ d$`total-site-count` <- NA }
+  if(is.null(d$`total-result-count`)){ d$`total-result-count` <- NA }
+  
+  df <- data.frame(characteristicType = char_type,
+                   siteType = site_type,
+                   startDate = start_date, 
+                   endDate = end_date,
+                   numSites = as.numeric(d$`total-site-count`),
+                   numResults = as.numeric(d$`total-result-count`),
+                   stringsAsFactors = FALSE)
+  return(df)
+}
+
+retryWQP <- function(..., retries=3){
+  
+  safeWQP = function(...){
+    result = tryCatch({
+      readWQPdata(...)
+    }, error = function(e) {
+      return(NULL)
+    })
+    return(result)
+  }
+  retry = 1
+  while (retry < retries){
+    result = safeWQP(...)
+    if (!is.null(result)){
+      retry = retries
+    } else {
+      message('query failed, retrying')
+      retry = retry+1
+    }
+  }
+  return(result)
+}
+
+tmp <- tempdir()
+
+counts_all_list <- lapply(char_types, dates_df = dates_df, tmp = tmp,
```

Comment:
So, I think there are two ways you could make this section more understandable and also more resilient to failure. First, you might get rid of the triple-nested apply by creating a master data frame with all permutations of inputs you want to run. Using something like `expand.grid` might quickly give you a master list that you can then just use an old-fashioned for loop on. 

Then, with such a for loop, you could use the data you are saving (line 83) more effectively as a cache. If a file with the results already exists, save yourself the time and don't run the request. Otherwise your saving of intermediate results isn't super useful as it can't be used to skip over results you already have. 

You don't necessarily have to get rid of the nested apply to use the cached result files, but I think it would make it cleaner.